### PR TITLE
Change navigation to be async

### PR DESCRIPTION
**Pull request by EBusch**

### CODE REVIEW

File: Float.Core/UX/INavigationContext.cs [hunk 1]
```diff
@@ -59,7 +60,7 @@ public interface INavigationContext
         /// Dismisses a modally-presented page.
         /// </summary>
         /// <param name="animated">Whether to animate the transition.</param>
-        void DismissPage(bool animated = true);
+        Task DismissPage(bool animated = true);
```
Comment (mediabounds) [inline Float.Core/UX/INavigationContext.cs:63]: If we make one of these awaitable, I'm inclined to make all methods in this interface awaitable. Thoughts?
(Also, the documentation would need to be updated here.)

File: Float.Core/UX/NavigationPageNavigationContext.cs [hunk 2]
```diff
@@ -93,13 +94,13 @@ public void PresentPage(Page page, bool animated = true)
         }
 
         /// <inheritdoc />
-        public void DismissPage(bool animated = true)
+        public async Task DismissPage(bool animated = true)
         {
-            DeviceProxy.BeginInvokeOnMainThread(() =>
+            await Device.InvokeOnMainThreadAsync(async () =>
```
Comment (mediabounds) [inline Float.Core/UX/NavigationPageNavigationContext.cs:99]: We would need to continue using the `DeviceProxy` class (see https://github.com/gowithfloat/Float.Core/commit/ec7d9f34c51f7e6f7ebc0f64617032cbd12edbba).

**Review by mediabounds — CHANGES_REQUESTED** [review]

File: Float.Core/UX/NavigationPageNavigationContext.cs [hunk 3]
```diff
@@ -92,7 +93,7 @@ public void PresentPage(Page page, bool animated = true)
             });
         }
 
-        /// <inheritdoc />
+        /// <inheritdoc/>
```
Comment (mediabounds) [inline Float.Core/UX/NavigationPageNavigationContext.cs:96]: @EBusch Unnecessary change

File: Float.Core/UX/INavigationContext.cs [hunk 4]
```diff
@@ -42,25 +43,55 @@ public interface INavigationContext
...
         /// </summary>
         /// <param name="animated">Whether to animate the transition.</param>
         void DismissPage(bool animated = true);
 
+        /// <summary>
+        /// Dismisses a modally-presented page.
+        /// </summary>
+        /// <param name="animated">Whether to animate the transition.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        Task DismissPageAsync(bool animated = true);
```
Comment (mediabounds) [inline Float.Core/UX/INavigationContext.cs:93]: @EBusch Thoughts on updating the existing methods instead of adding new ones?
One the one hand, I would almost always say we should add to a library's API and avoid changing it to remain backward compatible.
On the other hand, in this case, this change would require implementations to make updates anyway (and thus isn't backwards compatible).

**Review by mediabounds — DISMISSED** [review]

File: Float.Core/UX/INavigationContext.cs (hunk 4 above)
Comment (EBusch) [reply Float.Core/UX/INavigationContext.cs:93]: @bbeversdorf thoughts on this? I think I lean towards the approach that doesn't break anything, though perhaps we consider marking these other methods as obsolete and remove them in a future update?

File: Float.Core/UX/NavigationPageNavigationContext.cs [hunk 5]
```diff
@@ -104,6 +105,45 @@ public void DismissPage(bool animated = true)
             });
         }
 
+        /// <inheritdoc />
+        public async Task DismissPageAsync(bool animated = true)
+        {
+            if (HasModal)
```
Comment (bbeversdorf) [inline Float.Core/UX/NavigationPageNavigationContext.cs:111]: @EBusch I noticed this was inside `DeviceProxy.BeginInvokeOnMainThread` for the original method. Is this thread safe? (or does it not matter)?

File: Float.Core/UX/NavigationPageNavigationContext.cs [hunk 6]
```diff
@@ -104,6 +105,45 @@ public void DismissPage(bool animated = true)
...
+        /// <inheritdoc/>
+        public async Task PopPageAsync(bool animated = true)
+        {
+            await DeviceProxy.InvokeOnMainThreadAsync(async () =>
+            {
+                await navigationPage.PopAsync(animated);
+            });
+        }
+
+        /// <inheritdoc/>
```
Comment (bbeversdorf) [inline Float.Core/UX/NavigationPageNavigationContext.cs:138]: @EBusch I think all of these should be consistent. I see `/// <inheritdoc/>` and `/// <inheritdoc />`

Comment (EBusch) [reply Float.Core/UX/NavigationPageNavigationContext.cs:138]: @bbeversdorf yeah I think you're right. Annoyingly, Visual Studio has updated the default behavior of fixing these. I will make them consistent.

**Review by bbeversdorf — APPROVED** [review]
We talked through most of the changes. LGTM
